src/lib.rs

```diff
@@ -3,12 +3,205 @@
 //! It is based on the linked list implementation in [Rust-for-Linux][1].
 //!
 //! [1]: https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/linked_list.rs
+//!
+//! In more general use cases, shoud not use RawList directly,
+//! suggest use smart pointers of nodes and move ownership of smart
+//! pointers to List:
+//!
+//! ```
+//! use linked_list::{GetLinks, Links, List};
+//!
+//! type InnerType = usize;
+//!
+//! pub struct ExampleNode {
+//!     pub inner: InnerType,
+//!     links: Links<Self>,
+//! }
+//!
+//! impl GetLinks for ExampleNode {
+//!     type EntryType = Self;
+//!
+//!     fn get_links(t: &Self) -> &Links<Self> {
+//!         &t.links
+//!     }
+//! }
+//!
+//! impl ExampleNode {
+//!     fn new(inner: InnerType) -> Self {
+//!         Self {
+//!             inner,
+//!             links: Links::new()
+//!         }
+//!     }
+//!
+//!     fn inner(&self) -> &InnerType {
+//!         &self.inner
+//!     }
+//! }
+//!
+//! let node1 = Box::new(ExampleNode::new(0));
+//! let node2 = Box::new(ExampleNode::new(1));
+//! let mut list =  List::<Box<ExampleNode>>::new();
+//!
+//! list.push_back(node1);
+//! list.push_back(node2);
+//!
+//! //Support Iter
+//! for (i,e) in list.iter().enumerate() {
+//!     assert!(*e.inner() == i);
+//! }
+//!
+//! // Pop drop
+//! assert!(*list.pop_front().unwrap().inner() == 0);
+//! assert!(*list.pop_front().unwrap().inner() == 1);
+//!
+//! ```
+//!
 
-#![no_std]
+#![cfg_attr(not(test), no_std)]
 
 mod linked_list;
+mod raw_list;
+pub use linked_list::List;
+pub use raw_list::{GetLinks, Links};
+
+/// Defines a new node type that wraps an inner type and includes links for List.
+///
+/// # Parameters
+///
+/// - `struct_name`: The name of the struct to define.
+/// - `type`: The inner type to wrap.
+///
+/// # Example
+///
+/// ```rust
+/// use linked_list::{def_node, List};
+///
+/// def_node!(ExampleNode, usize);
+///
+/// let node1 = Box::new(ExampleNode::new(0));
+/// let node2 = Box::new(ExampleNode::new(1));
+/// let mut list =  List::<Box<ExampleNode>>::new();
+///
+/// list.push_back(node1);
+/// list.push_back(node2);
+///
+/// for (i,e) in list.iter().enumerate() {
+///     assert!(*e.inner() == i);
+/// }
+/// ```
+#[macro_export]
+macro_rules! def_node {
+    ($struct_name:ident, $type:ty) => {
+        #[doc = "A node wrapper for inner type "]
```

Comment:
How to deal with the user-defined doc of the node?

Reply:
I suggest this api style:
```rs
define_node! {
    /// A node with usize value.
    struct UsizeNode(usize);
    /// A node with inner type.
    struct WrapperNode<T>(T);
}

Reply:
fix： now we can use like this : 
```rust
def_node!(
     /// An example Node with usize
     struct ExampleNode1(usize);
     /// An example Node with generic Inner type
     struct GenericNode1<T>(T);
);
```
can create : 
```rust
/// An example Node with usize
struct ExampleNode1 {
    inner: usize,
    links: crate::Links<Self>,
}
impl crate::GetLinks for ExampleNode1 {
    type EntryType = Self;
    #[inline]
    fn get_links(t: &Self) -> &crate::Links<Self> {
        &t.links
    }
}
impl ExampleNode1 {
    ///Create a node
    const fn new(inner: usize) -> Self {
        Self {
            inner,
            links: crate::Links::new(),
        }
    }
    #[inline]
    ///Get inner
    const fn inner(&self) -> &usize {
        &self.inner
    }
}
impl core::ops::Deref for ExampleNode1 {
    type Target = usize;
    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
/// An example Node with generic Inner type
struct GenericNode1<T> {
    inner: T,
    links: crate::Links<Self>,
}
impl<T> crate::GetLinks for GenericNode1<T> {
    type EntryType = Self;
    #[inline]
    fn get_links(t: &Self) -> &crate::Links<Self> {
        &t.links
    }
}
impl<T> GenericNode1<T> {
    ///Create a node
    const fn new(inner: T) -> Self {
        Self {
            inner,
            links: crate::Links::new(),
        }
    }
    #[inline]
    ///Get inner
    const fn inner(&self) -> &T {
        &self.inner
    }
}
impl<T> core::ops::Deref for GenericNode1<T> {
    type Target = T;
    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
```